timemachine/cpp/src/local_md_potentials.cu

```diff
@@ -0,0 +1,191 @@
+    gpuErrchk(cudaPeekAtLastError());
+
+    this->_setup_free_idxs_given_reference_idx(reference_idx, radius, k, stream);
+}
+
+void LocalMDPotentials::_setup_free_idxs_given_reference_idx(
+    const unsigned int reference_idx, const double radius, const double k, const cudaStream_t stream) {
+    const int tpb = warp_size;
+
+    LessThan select_op(N_);
+
+    // Partition the free idxs into the row idxs
+    gpuErrchk(cub::DevicePartition::If(
+        d_temp_storage_buffer_->data,
+        temp_storage_bytes_,
+        d_free_idxs_.data,
+        d_row_idxs_.data,
+        num_selected_buffer_.data,
+        N_,
+        select_op,
+        stream));
+
+    gpuErrchk(cudaMemcpyAsync(
+        p_num_selected_.data,
+        num_selected_buffer_.data,
+        1 * sizeof(*p_num_selected_.data),
+        cudaMemcpyDeviceToHost,
+        stream));
+    gpuErrchk(cudaStreamSynchronize(stream));
+
+    const int num_row_idxs = p_num_selected_.data[0];
+    const int num_col_idxs = N_ - num_row_idxs;
+
+    if (num_row_idxs == 0) {
+        throw std::runtime_error("LocalMDPotentials setup has no free particles selected");
+    }
+
+    // The reference particle will always be in the column idxs
+    if (num_row_idxs == N_ - 1) {
+        fprintf(stderr, "LocalMDPotentials setup has entire system selected\n");
+    }
+
+    k_construct_bonded_params<<<ceil_divide(num_row_idxs, tpb), tpb, 0, stream>>>(
+        num_row_idxs, N_, reference_idx, k, 0.0, radius, d_row_idxs_.data, restraint_pairs_.data, bond_params_.data);
+    gpuErrchk(cudaPeekAtLastError());
+
+    // Setup the flat bottom restraints
+    bound_restraint_->set_params_device(std::vector<int>({num_row_idxs, 3}), bond_params_.data, stream);
+    restraint_->set_bonds_device(num_row_idxs, restraint_pairs_.data, stream);
+
+    // Set the nonbonded potential to compute forces of free particles
+    set_nonbonded_potential_idxs(nonbonded_bp_->potential, num_row_idxs, d_row_idxs_.data, stream);
+
+    // Invert to get column idxs
+    k_invert_indices<<<ceil_divide(N_, tpb), tpb, 0, stream>>>(N_, d_free_idxs_.data);
+    gpuErrchk(cudaPeekAtLastError());
+
+    // Partition the column idxs to the column buffer to setup the interaction group
+    gpuErrchk(cub::DevicePartition::If(
+        d_temp_storage_buffer_->data,
+        temp_storage_bytes_,
+        d_free_idxs_.data,
+        d_col_idxs_.data,
+        num_selected_buffer_.data,
+        N_,
+        select_op,
+        stream));
+
+    // Free particles should be in the row idxs
+    set_nonbonded_ixn_potential_idxs(
+        ixn_group_->potential, num_col_idxs, num_row_idxs, d_col_idxs_.data, d_row_idxs_.data, stream);
+}
+
+std::vector<std::shared_ptr<BoundPotential>> LocalMDPotentials::get_potentials() { return all_potentials_; }
+
+DeviceBuffer<unsigned int> *LocalMDPotentials::get_free_idxs() { return &d_row_idxs_; }
+
+void LocalMDPotentials::reset(const cudaStream_t stream) {
+    // Set the row idxs back to the identity.
```

Comment:
we don't need to do anything with `d_col_idxs_` or other buffers here?

Reply:
(usually `reset` semantically resets all the internal state, here it looks like we're just resetting the row_idxs)

Reply:
Will rename to `reset_potentials` as really it is here to reset the potentials that it had modified for local MD back to their original state. At the moment the only potential modified is the NonbondedAllPairs potential to run on the free idxs.

Reply:
Renamed and added doc string https://github.com/proteneer/timemachine/pull/1004/commits/ebdbd0e0581ac1b15718e1c3497869c7ae7aa3d4